### ISSUE TITLE
[discuss] : any suggestions to help me diagnose whether this result is as expected?

### ISSUE THREAD

**Issue by jensenojs** [cloudwego/goref#4]
As the title says, I suspect there is a memory leak in our database, so I want to diagnose it. but I don't see the stack trace for the database code

<img width="1742" alt="image" src="https://github.com/user-attachments/assets/dd7e4968-7fdf-4aa0-818d-a9297b475a61">

[grp_and_heap.zip](https://github.com/user-attachments/files/16188254/grp_and_heap.zip)

**Comment by jayantxie** [comment]
<img width="1920" alt="image" src="https://github.com/user-attachments/assets/42644330-6133-4678-9178-c786c36db3cc">

"finalized" indicates that an object has been set with a finalizer using `runtime.SetFinalizer`. Since the finalizer cannot determine the type, it cannot record the original type information. If there are a large number of finalized objects, it is possible that the finalizer objects have circular references. You could check if this is the case.

See https://github.com/golang/go/blob/master/src/runtime/mfinal.go#L341

**Comment by jensenojs** [comment]
Yes, we do use `runtime.SetFinalizer` in several places, for example, to close channels or do other detection mechanisms.

I didn't find a direct API (e.g. something like `runtime.NumberFinalizer`), but I can try to add logs to locate the problem, and it seems to be OK.

***

Previously we faced memory leaks, where the heap could only be seen as a leak, but not who it was referenced by. This tool would be useful for my work, and I'm also interested in program analysis (but I know nothing about this field right now), is there anything we can do to participate in the development of this library?

**Comment by jayantxie** [comment]
Welcome to participate in the development of the tool to make it more user-friendly. You can start by getting a general understanding of the principles and the overall implementation. I have two tasks for you to try, which will help you become more familiar with the development.
The first task is to enhance the current output of the node names. https://github.com/cloudwego/goref/issues/5
The second task is related to the pointer mask. https://github.com/cloudwego/goref/issues/6

**Comment by jensenojs** [comment]
Okay, I'll try it when I have time. 

I'll try it out when I have time. Also I'd like to suggest that we add some little examples like best practice or quick start 101, I realized that we do have a memory leak. I was using the library the wrong way, running it during runtime. 

If I wanted to see the call relationship of the memory leak, I should have waited for the load to run and let it idle for say an hour, and then grabbed the information.
heap : 
<img width="1649" alt="image" src="https://github.com/user-attachments/assets/7c8a37c6-f856-47c3-824a-9b47473abd48">

grp : 
<img width="1214" alt="image" src="https://github.com/user-attachments/assets/aed7eddb-b740-4aef-a580-4cf05fe674c7">

As you can see at this point there are indeed some call relationships that belong to our code, but I'm not sure if this information has anything to do with the memory leak.

**Comment by jayantxie** [comment]
According to the grf flame graph you provided, the issue primarily arises from memory management of finalized objects. If these objects cannot be released even after being idle for a long time, it is highly likely that `runtime.SetFinalizer` function has been misused.

**Comment by jayantxie** [comment]
> I'll try it out when I have time. Also I'd like to suggest that we add some little examples like best practice or quick start 101, I realized that we do have a memory leak.

Good idea, I'll mark it as TODO.
